### PR TITLE
[SyntaxParse] Parse PoundAssertStmt

### PR DESCRIPTION
**Pull request by nathawes**
Also added parseExpressionSyntax as an intermediate syntax parser until
expression parsing is completely implemented.

### CODE REVIEW

**Comment by nathawes** [comment]
@swift-ci please test

File: lib/Parse/ASTGen.cpp [hunk 1]
```diff
@@ -711,10 +711,50 @@ LayoutConstraint ASTGen::generate(const LayoutConstraintSyntax &constraint,
...
+Stmt *ASTGen::generate(const syntax::PoundAssertStmtSyntax &Stmt,
+                       const SourceLoc Loc) {
+  // Don't form a PoundAssertStmt without a condition
+  if (Stmt.getCondition().isUnknown())
+    return nullptr;
+  SourceLoc CondLoc = advanceLocBegin(Loc, Stmt.getCondition());
+  if (!hasExpr(CondLoc))
+    return nullptr;
+
+  Expr *CondExpr = getExpr(CondLoc);
```
Comment (rintaro) [inline lib/Parse/ASTGen.cpp]: Once we have `generate(ExprSyntax)`, this block should be like
```cpp
auto CondExpr = generate(Stmt.getCondition(), Loc);
if (! CondExpr)
  // Don't form a PoundAssertStmt without a condition.
  return nullptr;
```

File: lib/Syntax/Syntax.cpp [hunk 2]
```diff
@@ -98,6 +98,12 @@ llvm::Optional<Syntax> Syntax::getChild(const size_t N) const {
   return Syntax {Root, ChildData.get()};
 }
 
+Optional<Syntax> Syntax::getPreviousNode() const {
+  if (auto prev = getData().getPreviousNode())
+    return TokenSyntax(Root, prev.get());
```
Comment (rintaro) [inline lib/Syntax/Syntax.cpp]: I believe `Syntax` instead of `TokenSyntax`.

File: lib/Parse/ParseExpr.cpp [hunk 3]
```diff
@@ -34,6 +34,18 @@
 using namespace swift;
 using namespace swift::syntax;
 
+ParsedSyntaxResult<ParsedExprSyntax> Parser::parseExpressionSyntax(Diag<> ID) {
+  SourceLoc ExprLoc = Tok.getLoc();
+  SyntaxParsingContext ExprParsingContext(SyntaxContext, SyntaxContextKind::Expr);
```
Comment (rintaro) [inline lib/Parse/ParseExpr.cpp]: nitpick: 80 columns

File: lib/Parse/ParseStmt.cpp [hunk 4]
```diff
@@ -2529,49 +2531,76 @@ ParserResult<CaseStmt> Parser::parseStmtCase(bool IsActive) {
...
 
-  if (Tok.isNot(tok::l_paren)) {
-    diagnose(Tok, diag::pound_assert_expected_lparen);
-    return makeParserError();
+  SourceLoc startLoc = Tok.getLoc();
+  builder.usePoundAssert(consumeTokenSyntax(tok::pound_assert));
+  if (!Tok.isFollowingLParen()) {
+    diagnose(Tok.getLoc(), diag::pound_assert_expected_lparen);
+    status.setIsParseError();
+    builder.useCondition(ParsedSyntaxRecorder::makeUnknownExpr({}, *SyntaxContext));
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: 80 columns.

File: lib/Parse/ParseStmt.cpp [hunk 5]
```diff
@@ -2529,49 +2531,76 @@ ParserResult<CaseStmt> Parser::parseStmtCase(bool IsActive) {
...
+    return makeParsedResult(builder.build(), status);
   }
-  SourceLoc LBLoc = consumeToken(tok::l_paren);
+  SourceLoc lParenLoc = Tok.getLoc();
+  builder.useLeftParen(consumeTokenSyntax(tok::l_paren));
 
-  auto conditionExprResult = parseExpr(diag::pound_assert_expected_expression);
-  if (conditionExprResult.isParseError())
-    return ParserStatus(conditionExprResult);
+  auto conditionExprResult = parseExpressionSyntax(diag::pound_assert_expected_expression);
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: 80 columns

File: lib/Parse/ParseStmt.cpp [hunk 6]
```diff
@@ -2529,49 +2531,76 @@ ParserResult<CaseStmt> Parser::parseStmtCase(bool IsActive) {
...
-                                                        "'#assert' message");
-    consumeToken();
-    if (!messageOpt)
-      return makeParserError();
-
-    message = *messageOpt;
+    auto MessageLoc = Tok.getLoc();
+    // FIXME: Support extended escaping string literal.
+    if (Tok.getCustomDelimiterLen()) {
+      diagnose(MessageLoc, diag::forbidden_extended_escaping_string, "'#assert' message");
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: ditto

File: lib/Parse/ParseStmt.cpp [hunk 7]
```diff
@@ -2529,49 +2531,76 @@ ParserResult<CaseStmt> Parser::parseStmtCase(bool IsActive) {
...
+      status.setIsParseError();
+    } else {
+      SmallVector<Lexer::StringSegment, 1> Segments;
+      L->getStringLiteralSegments(Tok, Segments);
+      if (Segments.size() != 1 ||
+          Segments.front().Kind == Lexer::StringSegment::Expr) {
+        diagnose(MessageLoc, diag::forbidden_interpolated_string, "'#assert' message");
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: ditto

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Pull request closed without merging**